system/src/ledger/ledger_manager.cpp

```diff
@@ -0,0 +1,1599 @@
+        coap_remove_connection_handler(connectionCallback, nullptr);
+    }
+}
+
+int LedgerManager::init() {
+    std::lock_guard lock(mutex_);
+    if (state_ != State::NEW) {
+        return 0; // Already initialized
+    }
+    // Device must not be connected to the cloud
+    if (spark_cloud_flag_connected()) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    // TODO: Allow seeking in ledger and CoAP message streams so that an intermediate buffer is not
+    // needed when streaming ledger data to and from the server
+    std::unique_ptr<char[]> buf(new char[STREAM_BUFFER_SIZE]);
+    if (!buf) {
+        return SYSTEM_ERROR_NO_MEMORY;
+    }
+    // Enumerate local ledgers
+    LedgerSyncContexts contexts;
+    FsLock fs;
+    lfs_dir_t dir = {};
+    int r = lfs_dir_open(fs.instance(), &dir, LEDGER_ROOT_DIR);
+    if (r == 0) {
+        SCOPE_GUARD({
+            int r = closeDir(fs.instance(), &dir);
+            if (r < 0) {
+                LOG(ERROR, "Failed to close directory handle: %d", r);
+            }
+        });
+        lfs_info entry = {};
+        while ((r = lfs_dir_read(fs.instance(), &dir, &entry)) == 1) {
+            if (entry.type != LFS_TYPE_DIR) {
+                LOG(WARN, "Found unexpected entry in ledger directory");
+                continue;
+            }
+            if (std::strcmp(entry.name, ".") == 0 || std::strcmp(entry.name, "..") == 0) {
+                continue;
+            }
+            if (contexts.size() >= (int)MAX_LEDGER_COUNT) {
+                LOG(ERROR, "Maximum number of ledgers reached, skipping ledger: %s", entry.name);
+                continue;
+            }
+            // Load the ledger info
+            Ledger ledger;
+            int r = ledger.init(entry.name);
+            if (r < 0) {
+                LOG(ERROR, "Failed to initialize ledger: %d", r);
+                continue;
+            }
+            // Create a sync context for the ledger
+            std::unique_ptr<LedgerSyncContext> ctx(new(std::nothrow) LedgerSyncContext());
+            if (!ctx) {
+                return SYSTEM_ERROR_NO_MEMORY;
+            }
+            size_t n = strlcpy(ctx->name, entry.name, sizeof(ctx->name));
+            if (n >= sizeof(ctx->name)) {
+                return SYSTEM_ERROR_INTERNAL; // Name length is validated in Ledger::init()
+            }
+            auto info = ledger.info();
+            ctx->scopeId = info.scopeId();
+            ctx->syncDir = info.syncDirection();
+            ctx->syncPending = info.syncPending();
+            if (ctx->syncDir == LEDGER_SYNC_DIRECTION_CLOUD_TO_DEVICE) {
+                ctx->lastUpdated = info.lastUpdated();
+            }
+            // Context objects are sorted by ledger name
+            bool found = false;
+            auto it = findContext(contexts, ctx->name, found);
+            assert(!found);
+            it = contexts.insert(it, std::move(ctx));
+            if (it == contexts_.end()) {
```

Comment:
Is this supposed to be `contexts.end()`? This whole loop is working with a local collection of sync contexts.

Reply:
Good catch! This code used to populate the global `contexts_` previously.

Reply:
Fixed.